core/arch/riscv/include/mm/core_mmu_arch.h

```diff
@@ -67,25 +67,38 @@
 #define RISCV_PGLEVELS		((RISCV_MMU_VA_WIDTH - RISCV_PGSHIFT) / \
 							 RISCV_PGLEVEL_BITS)
 #define RISCV_MMU_VPN_MASK	(BIT(RISCV_PGLEVEL_BITS) - 1)
+#define RISCV_MMU_MAX_PGTS	16
 
 #define SMALL_PAGE_SHIFT	U(12)
 
 /*
- * Level 0: shift = 12 4K pages
- * Level 1: shift = 21 2M pages
- * Level 2: shift = 30 1G pages
+ * Level 0, shift = 12, 4 KiB pages
+ * Level 1, shift = 21, 2 MiB pages (4 MiB pages in Sv32)
+ * Level 2, shift = 30, 1 GiB pages
+ * Level 3, shift = 39, 512 GiB pages
+ * Level 4, shift = 48, 256 TiB pages
  */
+#define CORE_MMU_SHIFT_OF_LEVEL(level) (RISCV_PGLEVEL_BITS * \
+					(level) + \
+					RISCV_PGSHIFT)
 
 #define CORE_MMU_USER_CODE_SHIFT	SMALL_PAGE_SHIFT
 #define CORE_MMU_USER_PARAM_SHIFT	SMALL_PAGE_SHIFT
 
+/*
+ * In all MMU modes, the CORE_MMU_PGDIR_LEVEL is always 0:
+ * Sv32: 4 KiB, 4 MiB
+ * Sv39: 4 KiB, 2 MiB, 1 GiB
+ * Sv48: 4 KiB, 2 MiB, 1 GiB, 512 GiB
+ * Sv57: 4 KiB, 2 MiB, 1 GiB, 512 GiB, 256 TiB
+ */
 #define CORE_MMU_PGDIR_LEVEL		U(0)
-#define CORE_MMU_PGDIR_SHIFT		U(12)
+#define CORE_MMU_PGDIR_SHIFT \
+		CORE_MMU_SHIFT_OF_LEVEL(CORE_MMU_PGDIR_LEVEL)
```

Comment:
`CORE_MMU_PGDIR_SHIFT` should be the shifted bits of megapages.

```suggestion
		CORE_MMU_SHIFT_OF_LEVEL(CORE_MMU_PGDIR_LEVEL + 1)
```

Reply:
Shift of level 0 should be 12 not 21 ?

Reply:
I am also confused about it.
But according to ARM's definition, the `CORE_MMU_PGDIR_SHIFT` is not for 4K page, it is for megapage ?

```
#ifdef CFG_WITH_LPAE
#define CORE_MMU_PGDIR_SHIFT	U(21)
#define CORE_MMU_PGDIR_LEVEL	U(3)
#else
#define CORE_MMU_PGDIR_SHIFT	U(20)
#define CORE_MMU_PGDIR_LEVEL	U(2)
#endif
```

Reply:
Yes that was my understanding. Anyway I'll keep it as you suggested.